implementation/src/main/java/io/smallrye/graphql/execution/QueryCache.java

```diff
@@ -0,0 +1,102 @@
+package io.smallrye.graphql.execution;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.jboss.logging.Logger;
+
+import graphql.ExecutionInput;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.parameters.InstrumentationValidationParameters;
+import graphql.execution.preparsed.PreparsedDocumentEntry;
+import graphql.execution.preparsed.PreparsedDocumentProvider;
+import graphql.validation.ValidationError;
+
+public class QueryCache extends SimpleInstrumentation implements PreparsedDocumentProvider {
+    private static final Logger LOG = Logger.getLogger(QueryCache.class);
+    private static final int MAX_CACHE_SIZE = AccessController.doPrivileged((PrivilegedAction<Integer>) () -> {
+        return Integer.getInteger("io.smallrye.graphql.execution.queryCacheMaxSize", 2048);
+    });
+
+    private final LRUCache<String, PreparsedDocumentEntry> cache = new LRUCache<>(MAX_CACHE_SIZE);
+    private final Map<String, ExecutionFunction> functionCache = new HashMap<>();
+    private final ThreadLocal<ExecutionFunction> executionFunctionTL = new ThreadLocal<>();
+
+    @Override
+    public PreparsedDocumentEntry getDocument(ExecutionInput executionInput,
+            Function<ExecutionInput, PreparsedDocumentEntry> computeFunction) {
+        String query = executionInput.getQuery();
+        PreparsedDocumentEntry entry = cache.get(query);
+        if (entry == null) {
+            ExecutionFunction executionFunction = new ExecutionFunction(computeFunction, executionInput);
+            functionCache.putIfAbsent(query, executionFunction);
```

Comment:
entries are only added, but never queried or deleted?

Reply:
The functionCache was an attempt to avoid the thread local, but ultimately wasn't feasible.  I removed most of that attempt, but left this in there.  Thanks for catching it.  I've removed it now too.